hy/compiler.py

```diff
@@ -1421,6 +1443,168 @@ def compile_dict_comprehension(self, expr):
+                parts.append(['for',
+                              self._storeize(part, self.compile(part)),
+                              self.compile(control.pop(0))])
+
+        # Get the `else`, if there is one.
+        orel = []
+        if is_for and ends_with_else(expr):
+            orel = [Result()]
+            for else_body in expr.pop()[1:]:
+                orel[0] += self.compile(else_body)
+                orel[0] += orel[0].expr_as_stmt()
+
+        # Produce a result.
+
+        if (is_for or elt.stmts or (key is not None and key.stmts) or
+            any(p[0] == 'expr' or (p[2].stmts if p[0] in ("for", "setv") else p[1].stmts)
+                for p in parts)):
+            # The desired comprehension can't be expressed as a
+            # real Python comprehension. We'll write it as a nested
+            # loop instead.
+            contains_yield = []
+            def f(parts):
+                # This function is called recursively to construct
+                # the nested loop.
+                if not parts:
+                    if is_for:
+                        if expr:
+                            body = self._compile_branch(expr)
+                            if body.contains_yield:
+                                contains_yield.append(True)
+                            return body + body.expr_as_stmt()
+                        return Result(stmts=[asty.Pass(expr)])
+                    if node_class is asty.DictComp:
+                        ret = key + elt
+                        val = asty.Tuple(
+                            key, ctx=ast.Load(),
+                            elts=[key.force_expr, elt.force_expr])
+                    else:
+                        ret = elt
+                        val = elt.force_expr
+                    return ret + asty.Expr(
+                        elt, value=asty.Yield(elt, value=val))
+                p, parts = parts[0], parts[1:]
+                if p[0] == "for":
+                    orelse = orel and orel.pop().stmts
+                    return p[2] + asty.For(
+                        p[2], target=p[1], iter=p[2].force_expr, body=f(parts).stmts,
+                        orelse=orelse)
+                elif p[0] == "setv":
+                    return p[2] + asty.Assign(
+                        p[1], targets=[p[1]], value=p[2].force_expr) + f(parts)
+                elif p[0] == "if":
+                    return p[1] + asty.If(
+                        p[1], test=p[1].force_expr, body=f(parts).stmts, orelse=[])
+                elif p[0] == "expr":
+                    return p[1] + p[1].expr_as_stmt() + f(parts)
+                else:
+                    raise ValueError("can't happen")
+            if is_for:
+                ret = f(parts)
+                ret.contains_yield = bool(contains_yield)
+                return ret
+            fname = self.get_anon_var()
+            ret = Result() + asty.FunctionDef(
```

Comment:
Won't this possibly cause some NameErrors? I see you do this in examples:

```clojure
(lfor x "abc" :do (setv y 1) x))
```

Is `y` expected to be accessible to the outer scope? In Python it would be, but because of the function call added here, it won't be.

Reply:
None of the three types of comprehensions, nor generator expressions, leak their variables into the surrounding scope in Python 2 and Python 3, with one exception: list comprehensions in Python 2. This seems more like an artifact than a feature, so I didn't try to retain it.

Reply:
Right, I see you point. The `for` and emitted ast threw me off. Thanks.